File: matlab/matrix-free-tensor/stokes_flow/bbfunplot.m
function w = bbfunplot(u, xy, xyp, mv, bbxe, bbye)

x = xy(:, 1);
xp = xyp(:, 1);
nvtx = length(x);
nu = 2 * nvtx;
np = 3 * length(xp);
nel = length(mv(:, 1));

ux = u(1:nvtx); 
uy = u(nvtx + 1:nu);

bbxes = squeeze(bbxe(1, :, :));
bbyes = squeeze(bbye(1, :, :));

w = zeros(nvtx, 1); 
for e = 1:nel
    ind = mv(e, :);
    wex = bbxes * uy(ind);
    wey = bbyes * ux(ind);
    w(ind) = w(ind) + wey;
    w(ind) = w(ind) - wex;
end

end

File: BrinkmanStokes/src/matrix-free-stacked/stokes_flow/bbfunplot.jl
function bbfunplot(u, kparams)

    xy = kparams["xy"]; xyp = kparams["xyp"]; mv = kparams["mv"]; bound = kparams["bound"]
    bbxe = kparams["bbxe"]; bbye = kparams["bbye"]

    # get variables
    nvtx = length(xy[:, 1]); nu = 2nvtx; np = 3length(xyp[:, 1])
    nel = length(mv[:, 1])
    bbxes = squeeze(bbxe[1, :, :], 1)
    bbyes = squeeze(bbye[1, :, :], 1)
    w = zeros(nvtx)
    # w = SharedArray(Float64, (nvtx, 1), pids = workers())

    ux = u[1:nvtx]
    uy = u[(nvtx + 1):nu]

    # @parallel for e = 1:nel
    for e in 1:nel
        ind = vec(mv[e, :]')
        w[ind] += bbxes * uy[ind]
        w[ind] -= bbyes * ux[ind]
    end

    return vec(w)
end
